File: apps/api/domain/use_cases/update_campaign_status.py
"""Use case: activate / pause / archive a campaign on the platform and in storage."""
from __future__ import annotations

from domain.models.campaign import Campaign, CampaignStatus
from domain.ports.outbound.ad_platform import IAdPlatform
from domain.ports.outbound.campaign_repository import ICampaignRepository


class UpdateCampaignStatusUseCase:
    def __init__(self, ad_platform: IAdPlatform, campaign_repo: ICampaignRepository):
        self.ad_platform = ad_platform
        self.campaign_repo = campaign_repo

    async def execute(self, campaign_id: str, status: str) -> Campaign:
        new_status = CampaignStatus(status)
        campaign = await self.campaign_repo.get(campaign_id)
        if campaign is None:
            raise LookupError(f"Campaign {campaign_id} not found")

        if campaign.meta_campaign_id:
            await self.ad_platform.update_status(
                "campaign", campaign.meta_campaign_id, new_status
            )

        updated = await self.campaign_repo.update_status(campaign_id, new_status)
        assert updated is not None
        return updated

File: apps/api-go/internal/domain/usecase/update_campaign_status.go
package usecase

import (
	"context"
	"fmt"

	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/model"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
)

// UpdateCampaignStatusUseCase activates, pauses or archives a campaign both on
// the ad platform and in our store, keeping the two in sync.
type UpdateCampaignStatusUseCase struct {
	adPlatform port.AdPlatform
	repo       port.CampaignRepository
}

func NewUpdateCampaignStatusUseCase(adPlatform port.AdPlatform, repo port.CampaignRepository) *UpdateCampaignStatusUseCase {
	return &UpdateCampaignStatusUseCase{adPlatform: adPlatform, repo: repo}
}

// Execute pushes the new status to the platform first, then records it locally.
// If the campaign is unknown the repository returns ErrNotFound, which the HTTP
// layer maps to 404.
func (uc *UpdateCampaignStatusUseCase) Execute(ctx context.Context, campaignID string, status model.CampaignStatus) (model.Campaign, error) {
	campaign, err := uc.repo.Get(ctx, campaignID)
	if err != nil {
		return model.Campaign{}, err
	}
	if campaign.MetaCampaignID != "" {
		if err := uc.adPlatform.UpdateStatus(ctx, "campaign", campaign.MetaCampaignID, status); err != nil {
			return model.Campaign{}, fmt.Errorf("platform update status: %w", err)
		}
	}
	return uc.repo.UpdateStatus(ctx, campaignID, status)
}
